File: impl/python/weavepack_log/decoder.py
"""weavepack-log — pure-Python decoder (event batches + stream headers + delta chains).

Profile isolation: imports only from .types. No JSON/tensor/wire/tabular profile code.
"""

import struct
from .types import (
    CTYPE, OP, SCHEMA_SUB_OP,
    FRAME_SNAPSHOT, FRAME_DELTA, FRAME_STREAM_HEADER,
    SCHEMA_HASH_BYTES, STREAM_ID_BYTES, MAX_STRING_BYTES,
    null_bitmap_bytes, get_null_bit,
)


class _ByteReader:
    def __init__(self, buf: bytes):
        self._buf = bytes(buf)
        self._pos = 0

    def eof(self) -> bool:
        return self._pos >= len(self._buf)

    def read_byte(self) -> int:
        if self._pos >= len(self._buf):
            raise ValueError("unexpected end of input")
        b = self._buf[self._pos]
        self._pos += 1
        return b

    def read_bytes(self, n: int) -> bytes:
        if self._pos + n > len(self._buf):
            raise ValueError("unexpected end of input")
        out = self._buf[self._pos:self._pos + n]
        self._pos += n
        return out

    def read_leb128(self) -> int:
        result = 0
        shift = 0
        while True:
            b = self.read_byte()
            result |= (b & 0x7F) << shift
            shift += 7
            if (b & 0x80) == 0:
                break
            if shift >= 35:
                raise ValueError("LEB128 overflow for uint32")
        return result & 0xFFFFFFFF

    def read_leb128_big(self) -> int:
        result = 0
        shift = 0
        while True:
            b = self.read_byte()
            result |= (b & 0x7F) << shift
            shift += 7
            if (b & 0x80) == 0:
                break
            if shift >= 70:
                raise ValueError("LEB128 overflow for uint64")
        return result

    def read_zigzag64(self) -> int:
        enc = self.read_leb128_big()
        v = (enc >> 1) ^ -(enc & 1)
        if v >= (1 << 63):
            v -= (1 << 64)
        return v


def _read_seq_block(r: _ByteReader, num_events: int) -> list:
    if num_events == 0:
        return []
    first = r.read_leb128_big()
    seqs = [first]
    prev = first
    for _ in range(1, num_events):
        delta = r.read_leb128_big()
        if delta < 1:
            raise ValueError("duplicate_seq: seq delta must be >=1")
        prev = prev + delta
        seqs.append(prev)
    return seqs


def _read_ts_block(r: _ByteReader, num_events: int) -> list:
    if num_events == 0:
        return []
    first = r.read_zigzag64()
    tss = [first]
    prev = first
    for _ in range(1, num_events):
        delta = r.read_leb128_big()
        prev = prev + delta
        tss.append(prev)
    return tss


def _read_value(r: _ByteReader, ctype: int):
    if ctype == CTYPE.BOOL:
        return r.read_byte() != 0
    elif ctype == CTYPE.INT8:
        b = r.read_byte()
        return b - 256 if b > 127 else b
    elif ctype == CTYPE.INT16:
        raw = r.read_bytes(2)
        u = raw[0] | (raw[1] << 8)
        return u - 0x10000 if u > 0x7FFF else u
    elif ctype == CTYPE.INT32:
        raw = r.read_bytes(4)
        u = struct.unpack("<I", raw)[0]
        return u - 0x100000000 if u > 0x7FFFFFFF else u
    elif ctype == CTYPE.INT64:
        raw = r.read_bytes(8)
        lo, hi = struct.unpack("<II", raw)
        u = lo | (hi << 32)
        return u - (1 << 64) if u > 0x7FFFFFFFFFFFFFFF else u
    elif ctype == CTYPE.UINT8:
        return r.read_byte()
    elif ctype == CTYPE.UINT16:
        raw = r.read_bytes(2)
        return raw[0] | (raw[1] << 8)
    elif ctype == CTYPE.UINT32:
        raw = r.read_bytes(4)
        return struct.unpack("<I", raw)[0]
    elif ctype == CTYPE.UINT64:
        raw = r.read_bytes(8)
        lo, hi = struct.unpack("<II", raw)
        return lo | (hi << 32)
    elif ctype == CTYPE.FLOAT32:
        raw = r.read_bytes(4)
        return struct.unpack("<f", raw)[0]
    elif ctype == CTYPE.FLOAT64:
        raw = r.read_bytes(8)
        return struct.unpack("<d", raw)[0]
    elif ctype == CTYPE.STRING:
        length = r.read_leb128()
        if length > MAX_STRING_BYTES:
            raise ValueError("string exceeds 256 MiB limit")
        raw = r.read_bytes(length)
        try:
            return raw.decode("utf-8")
        except UnicodeDecodeError:
            raise ValueError("invalid_utf8: string column contains invalid UTF-8")
    elif ctype == CTYPE.BYTES:
        length = r.read_leb128()
        if length > MAX_STRING_BYTES:
            raise ValueError("bytes exceeds 256 MiB limit")
        return r.read_bytes(length)
    elif ctype == CTYPE.DATE32:
        raw = r.read_bytes(4)
        u = struct.unpack("<I", raw)[0]
        return u - 0x100000000 if u > 0x7FFFFFFF else u
    elif ctype == CTYPE.TIMESTAMP64:
        raw = r.read_bytes(8)
        lo, hi = struct.unpack("<II", raw)
        u = lo | (hi << 32)
        return u - (1 << 64) if u > 0x7FFFFFFFFFFFFFFF else u
    elif ctype == CTYPE.LEVEL:
        v = r.read_byte()
        if v > 5:
            raise ValueError(f"unknown_level: level value {v} is reserved")
        return v
    else:
        raise ValueError(f"unknown_ctype: ctype {ctype}")


def _read_bool_column(r: _ByteReader, count: int) -> list:
    n_bytes = (count + 7) // 8
    raw = r.read_bytes(n_bytes)
    return [bool((raw[i >> 3] >> (7 - (i & 7))) & 1) for i in range(count)]


def _read_level_column(r: _ByteReader, count: int) -> list:
    n_bytes = (count * 3 + 7) // 8
    raw = r.read_bytes(n_bytes)
    values = []
    for i in range(count):
        bit_base = i * 3
        v = 0
        for b in range(3):
            pos = bit_base + b
            if (raw[pos >> 3] >> (pos & 7)) & 1:
                v |= (1 << b)
        if v >= 6:
            raise ValueError(f"unknown_level: level value {v} is reserved")
        values.append(v)
    total_bits = count * 3
    used_in_last = total_bits & 7
    if used_in_last != 0 and n_bytes > 0:
        last = raw[n_bytes - 1]
        mask = 0xFF << used_in_last
        if last & mask & 0xFF:
            raise ValueError("invalid_level_padding: padding bits must be zero")
    return values


def _read_null_bitmap(r: _ByteReader, num_events: int) -> list:
    n_bytes = null_bitmap_bytes(num_events)
    raw = r.read_bytes(n_bytes)
    rem = num_events & 7
    if rem != 0:
        last = raw[n_bytes - 1]
        mask = 0xFF >> rem
        if last & mask:
            raise ValueError("invalid_null_bitmap: padding bits must be zero")
    return [get_null_bit(raw, i) for i in range(num_events)]


def _read_value_column(r: _ByteReader, ctype: int, count: int) -> list:
    if ctype == CTYPE.BOOL:
        return _read_bool_column(r, count)
    elif ctype == CTYPE.LEVEL:
        return _read_level_column(r, count)
    else:
        return [_read_value(r, ctype) for _ in range(count)]


def _read_column_block(r: _ByteReader, num_events: int) -> dict:
    col_id = r.read_leb128()
    if col_id < 2:
        raise ValueError(f"reserved_col_id: col_id {col_id} is reserved (must be >= 2)")
    type_byte = r.read_byte()
    ctype    = type_byte & 0x1F
    nullable = ((type_byte >> 5) & 1) == 1
    if ctype > 16:
        raise ValueError(f"unknown_ctype: ctype {ctype} >= 17 is reserved")

    if nullable:
        null_flags = _read_null_bitmap(r, num_events)
        non_null_count = sum(1 for b in null_flags if not b)
        raw_vals = _read_value_column(r, ctype, non_null_count)
        values = []
        vi = 0
        for is_null in null_flags:
            if is_null:
                values.append(None)
            else:
                values.append(raw_vals[vi])
                vi += 1
    else:
        values = _read_value_column(r, ctype, num_events)

    return {"colId": col_id, "ctype": ctype, "nullable": nullable, "values": values}


# ── Public: decode event batch (snapshot frame) ──────────────────────────────────────────


def decode_batch(data: bytes) -> dict:
    """Decode a snapshot frame (event batch).

    Returns { 'schemaHash': bytes, 'seqs': list[int], 'tss': list[int], 'columns': [...] }.
    """
    r = _ByteReader(data)
    flag = r.read_byte()
    if flag == FRAME_DELTA:
        raise ValueError("expected event batch (0x00), got delta chain (0x01)")
    if flag == FRAME_STREAM_HEADER:
        raise ValueError("expected event batch (0x00), got stream header (0x02)")
    if flag != FRAME_SNAPSHOT:
        raise ValueError(f"unknown frame flag 0x{flag:02x}")

    schema_hash = r.read_bytes(SCHEMA_HASH_BYTES)
    num_events  = r.read_leb128_big()

    seqs = _read_seq_block(r, num_events)
    tss  = _read_ts_block(r, num_events)

    num_user_cols = r.read_leb128()
    columns = [_read_column_block(r, num_events) for _ in range(num_user_cols)]

    return {"schemaHash": schema_hash, "seqs": seqs, "tss": tss, "columns": columns}


# ── Public: decode stream header ───────────────────────────────────────────────────────


def decode_stream_header(data: bytes) -> dict:
    """Decode a stream header frame.

    Returns { 'streamId': bytes, 'source': str, 'schemaHash': bytes, 'seqStart': int }.
    """
    r = _ByteReader(data)
    flag = r.read_byte()
    if flag != FRAME_STREAM_HEADER:
        raise ValueError(f"expected stream header (0x02), got 0x{flag:02x}")

    stream_id = r.read_bytes(STREAM_ID_BYTES)

    source_len   = r.read_leb128()
    source_bytes = r.read_bytes(source_len)
    try:
        source = source_bytes.decode("utf-8")
    except UnicodeDecodeError:
        raise ValueError("invalid_utf8: source contains invalid UTF-8")

    schema_hash = r.read_bytes(SCHEMA_HASH_BYTES)
    seq_start   = r.read_leb128_big()

    return {"streamId": stream_id, "source": source, "schemaHash": schema_hash, "seqStart": seq_start}


# ── Op decoding ──────────────────────────────────────────────────────────────────────


def _read_event_append(r: _ByteReader) -> dict:
    num_events = r.read_leb128_big()
    seqs = _read_seq_block(r, num_events)
    tss  = _read_ts_block(r, num_events)
    num_cols = r.read_leb128()
    columns = [_read_column_block(r, num_events) for _ in range(num_cols)]
    return {"op": OP.EVENT_APPEND, "seqs": seqs, "tss": tss, "columns": columns}


def _read_field_update(r: _ByteReader) -> dict:
    seq = r.read_leb128_big()
    num_cols = r.read_leb128()
    columns = []
    for _ in range(num_cols):
        col_id    = r.read_leb128()
        type_byte = r.read_byte()
        ctype     = type_byte & 0x1F
        has_value = ((type_byte >> 5) & 1) == 1
        if ctype > 16:
            raise ValueError(f"unknown_ctype: ctype {ctype} >= 17 is reserved")
        value = _read_value(r, ctype) if has_value else None
        columns.append({"colId": col_id, "ctype": ctype, "hasValue": has_value, "value": value})
    return {"op": OP.FIELD_UPDATE, "seq": seq, "columns": columns}


def _read_event_expire(r: _ByteReader) -> dict:
    seq_lo = r.read_leb128_big()
    seq_hi = r.read_leb128_big()
    if seq_lo > seq_hi:
        raise ValueError(f"invalid_seq_range: seq_lo ({seq_lo}) > seq_hi ({seq_hi})")
    return {"op": OP.EVENT_EXPIRE, "seqLo": seq_lo, "seqHi": seq_hi}


def _read_schema_evolve(r: _ByteReader) -> dict:
    sub_op = r.read_byte()
    if sub_op > 2:
        raise ValueError(f"unknown_schema_sub_op: sub_op {sub_op} is reserved")
    if sub_op == SCHEMA_SUB_OP.COLUMN_ADD:
        col_id    = r.read_leb128()
        type_byte = r.read_byte()
        ctype     = type_byte & 0x1F
        nullable  = ((type_byte >> 5) & 1) == 1
        if ctype > 16:
            raise ValueError(f"unknown_ctype: ctype {ctype} >= 17 is reserved")
        name_len = r.read_leb128()
        if name_len == 0:
            raise ValueError("invalid_col_name: empty name")
        name_bytes = r.read_bytes(name_len)
        name = name_bytes.decode("utf-8")
        return {"op": OP.SCHEMA_EVOLVE, "subOp": sub_op, "colId": col_id,
                "ctype": ctype, "nullable": nullable, "name": name}
    elif sub_op == SCHEMA_SUB_OP.COLUMN_DROP:
        col_id = r.read_leb128()
        return {"op": OP.SCHEMA_EVOLVE, "subOp": sub_op, "colId": col_id}
    else:  # COLUMN_RENAME
        col_id   = r.read_leb128()
        name_len = r.read_leb128()
        if name_len == 0:
            raise ValueError("invalid_col_name: empty name")
        name_bytes = r.read_bytes(name_len)
        name = name_bytes.decode("utf-8")
        return {"op": OP.SCHEMA_EVOLVE, "subOp": sub_op, "colId": col_id, "name": name}


def _read_cursor_checkpoint(r: _ByteReader) -> dict:
    seq      = r.read_leb128_big()
    name_len = r.read_leb128()
    if name_len == 0:
        raise ValueError("invalid_cursor_name: empty cursor name")
    name_bytes = r.read_bytes(name_len)
    name = name_bytes.decode("utf-8")
    return {"op": OP.CURSOR_CHECKPOINT, "seq": seq, "name": name}


def _read_op(r: _ByteReader) -> dict:
    op_code = r.read_byte()
    if op_code == OP.EVENT_APPEND:
        return _read_event_append(r)
    elif op_code == OP.FIELD_UPDATE:
        return _read_field_update(r)
    elif op_code == OP.EVENT_EXPIRE:
        return _read_event_expire(r)
    elif op_code == OP.SCHEMA_EVOLVE:
        return _read_schema_evolve(r)
    elif op_code == OP.CURSOR_CHECKPOINT:
        return _read_cursor_checkpoint(r)
    else:
        raise ValueError(f"unknown_delta_op: op code {op_code} is reserved")


# ── Public: decode delta chain ─────────────────────────────────────────────────────────────


def decode_chain(data: bytes) -> dict:
    """Decode a delta chain frame.

    Returns { 'schemaHash': bytes, 'ops': list }.
    """
    r = _ByteReader(data)
    flag = r.read_byte()
    if flag != FRAME_DELTA:
        raise ValueError(f"expected delta chain (0x01), got 0x{flag:02x}")

    schema_hash = r.read_bytes(SCHEMA_HASH_BYTES)
    num_ops = r.read_leb128()
    ops = [_read_op(r) for _ in range(num_ops)]

    return {"schemaHash": schema_hash, "ops": ops}

File: sdk/src/profiles/log/decoder.js
// weavepack-log — decoder (event batches + delta chains + stream header).
//
// Profile isolation: imports only from ./types.js. No JSON/tensor/wire/tabular code.

import {
  CTYPE, OP, SCHEMA_SUB_OP, FRAME_SNAPSHOT, FRAME_DELTA, FRAME_STREAM_HEADER,
  SCHEMA_HASH_BYTES, STREAM_ID_BYTES, MAX_STRING_BYTES,
  nullBitmapBytes, getNullBit,
} from "./types.js"

const _dec = new TextDecoder()

// ── ByteReader ─────────────────────────────────────────────────────────────

class ByteReader {
  constructor(bytes) {
    this._buf = bytes instanceof Uint8Array ? bytes : new Uint8Array(bytes)
    this._pos = 0
  }

  eof() { return this._pos >= this._buf.length }

  readByte() {
    if (this._pos >= this._buf.length) throw new Error("unexpected end of input")
    return this._buf[this._pos++]
  }

  readBytes(n) {
    if (this._pos + n > this._buf.length) throw new Error("unexpected end of input")
    const out = this._buf.slice(this._pos, this._pos + n)
    this._pos += n
    return out
  }

  readLEB128() {
    let result = 0, shift = 0
    while (true) {
      const b = this.readByte()
      result |= (b & 0x7F) << shift
      shift += 7
      if ((b & 0x80) === 0) break
      if (shift >= 35) throw new Error("LEB128 overflow for uint32")
    }
    return result >>> 0
  }

  readLEB128Big() {
    let result = 0n, shift = 0n
    while (true) {
      const b = BigInt(this.readByte())
      result |= (b & 0x7Fn) << shift
      shift += 7n
      if ((b & 0x80n) === 0n) break
      if (shift >= 70n) throw new Error("LEB128 overflow for uint64")
    }
    return result
  }

  // Read a zigzag-encoded sint64 (LEB128 uint64 then decode).
  readZigzag64() {
    const enc = this.readLEB128Big()
    return (enc >> 1n) ^ -(enc & 1n)
  }
}

// ── Level column decoding (3-bit LSB-first) ────────────────────────────────

function readLevelColumn(r, count) {
  const nBytes = Math.ceil(count * 3 / 8)
  const bytes = r.readBytes(nBytes)
  const values = []
  for (let i = 0; i < count; i++) {
    const bitBase = i * 3
    let v = 0
    for (let b = 0; b < 3; b++) {
      const pos = bitBase + b
      if ((bytes[pos >> 3] >> (pos & 7)) & 1) v |= (1 << b)
    }
    if (v >= 6) throw new Error(`unknown_level: level value ${v} is reserved`)
    values.push(v)
  }
  // Validate padding bits in final byte are zero.
  const totalBits = count * 3
  const usedInLastByte = totalBits & 7
  if (usedInLastByte !== 0) {
    const lastByte = bytes[nBytes - 1]
    const mask = 0xFF << usedInLastByte
    if (lastByte & mask & 0xFF) throw new Error("invalid_level_padding: padding bits must be zero")
  }
  return values
}

// ── Per-ctype single value decoding ───────────────────────────────────────

function readValue(r, ctype) {
  switch (ctype) {
    case CTYPE.BOOL:
      return r.readByte() !== 0
    case CTYPE.INT8: {
      const b = r.readByte()
      return b > 127 ? b - 256 : b
    }
    case CTYPE.INT16: {
      const bytes = r.readBytes(2)
      const u = bytes[0] | (bytes[1] << 8)
      return u > 0x7FFF ? u - 0x10000 : u
    }
    case CTYPE.INT32: {
      const bytes = r.readBytes(4)
      const u = (bytes[0] | (bytes[1] << 8) | (bytes[2] << 16) | (bytes[3] << 24)) >>> 0
      return u > 0x7FFFFFFF ? u - 0x100000000 : u
    }
    case CTYPE.INT64:
    case CTYPE.TIMESTAMP64: {
      const bytes = r.readBytes(8)
      const dv = new DataView(bytes.buffer, bytes.byteOffset, 8)
      const lo = BigInt(dv.getUint32(0, true))
      const hi = BigInt(dv.getUint32(4, true))
      const u = (hi << 32n) | lo
      return u > 0x7FFFFFFFFFFFFFFFn ? u - (1n << 64n) : u
    }
    case CTYPE.UINT8:
      return r.readByte()
    case CTYPE.UINT16: {
      const bytes = r.readBytes(2)
      return bytes[0] | (bytes[1] << 8)
    }
    case CTYPE.UINT32: {
      const bytes = r.readBytes(4)
      return (bytes[0] | (bytes[1] << 8) | (bytes[2] << 16) | (bytes[3] << 24)) >>> 0
    }
    case CTYPE.UINT64: {
      const bytes = r.readBytes(8)
      const dv = new DataView(bytes.buffer, bytes.byteOffset, 8)
      const lo = BigInt(dv.getUint32(0, true))
      const hi = BigInt(dv.getUint32(4, true))
      return (hi << 32n) | lo
    }
    case CTYPE.FLOAT32: {
      const bytes = r.readBytes(4)
      return new DataView(bytes.buffer, bytes.byteOffset, 4).getFloat32(0, true)
    }
    case CTYPE.FLOAT64: {
      const bytes = r.readBytes(8)
      return new DataView(bytes.buffer, bytes.byteOffset, 8).getFloat64(0, true)
    }
    case CTYPE.STRING: {
      const len = r.readLEB128()
      if (len > MAX_STRING_BYTES) throw new Error("string exceeds 256 MiB limit")
      const bytes = r.readBytes(len)
      try { return _dec.decode(bytes) }
      catch { throw new Error("invalid_utf8: string column contains invalid UTF-8") }
    }
    case CTYPE.BYTES: {
      const len = r.readLEB128()
      if (len > MAX_STRING_BYTES) throw new Error("bytes exceeds 256 MiB limit")
      return r.readBytes(len)
    }
    case CTYPE.DATE32: {
      const bytes = r.readBytes(4)
      const u = (bytes[0] | (bytes[1] << 8) | (bytes[2] << 16) | (bytes[3] << 24)) >>> 0
      return u > 0x7FFFFFFF ? u - 0x100000000 : u
    }
    case CTYPE.EXT:
      throw new Error("unknown_ext_type: EXT ctype not implemented in v0.1")
    case CTYPE.LEVEL:
      // Single level value (used in field_update): 1 byte.
      return r.readByte()
    default:
      throw new Error(`unknown_ctype: ctype ${ctype}`)
  }
}

// Read bool column: N values, MSB-first packed.
function readBoolColumn(r, count) {
  const nBytes = Math.ceil(count / 8)
  const bytes = r.readBytes(nBytes)
  const values = []
  for (let i = 0; i < count; i++) {
    values.push(((bytes[i >> 3] >> (7 - (i & 7))) & 1) === 1)
  }
  return values
}

// Read value column.
function readValueColumn(r, ctype, count) {
  if (ctype === CTYPE.BOOL)  return readBoolColumn(r, count)
  if (ctype === CTYPE.LEVEL) return readLevelColumn(r, count)
  const values = []
  for (let i = 0; i < count; i++) values.push(readValue(r, ctype))
  return values
}

// Read null bitmap and reconstruct the full values array.
function readNullBitmap(r, numEvents) {
  const nBytes = nullBitmapBytes(numEvents)
  const bytes = r.readBytes(nBytes)

  const rem = numEvents & 7
  if (rem !== 0) {
    const lastByte = bytes[nBytes - 1]
    const mask = 0xFF >> rem
    if (lastByte & mask) throw new Error("invalid_null_bitmap: padding bits must be zero")
  }

  const nulls = []
  for (let i = 0; i < numEvents; i++) {
    nulls.push(getNullBit(bytes, i) === 1)
  }
  return nulls
}

// ── Column block (5-bit ctype, 1-bit nullable) ─────────────────────────────

function readColumnBlock(r, numEvents) {
  const colId = r.readLEB128()
  if (colId < 2) throw new Error(`reserved_col_id: col_id ${colId} is reserved (must be ≥ 2)`)
  const typeByte = r.readByte()
  const ctype    = typeByte & 0x1F
  const nullable = ((typeByte >> 5) & 1) === 1
  if (ctype > 16) throw new Error(`unknown_ctype: ctype ${ctype} ≥ 17 is reserved`)

  let values
  if (nullable) {
    const nullFlags = readNullBitmap(r, numEvents)
    const nonNullCount = nullFlags.filter(b => !b).length
    const nonNullValues = readValueColumn(r, ctype, nonNullCount)
    values = []
    let vi = 0
    for (let i = 0; i < numEvents; i++) {
      values.push(nullFlags[i] ? null : nonNullValues[vi++])
    }
  } else {
    values = readValueColumn(r, ctype, numEvents)
  }

  return { colId, ctype, nullable, values }
}

// ── Mandatory seq block ────────────────────────────────────────────────────

function readSeqBlock(r, numEvents) {
  if (numEvents === 0) return []
  const seqs = []
  const firstSeq = r.readLEB128Big()
  seqs.push(firstSeq)
  let prev = firstSeq
  for (let i = 1; i < numEvents; i++) {
    const delta = r.readLEB128Big()
    if (delta < 1n) throw new Error("duplicate_seq: seq delta must be ≥1")
    prev = prev + delta
    seqs.push(prev)
  }
  return seqs
}

// ── Mandatory ts block ─────────────────────────────────────────────────────

function readTsBlock(r, numEvents) {
  if (numEvents === 0) return []
  const tss = []
  const firstTs = r.readZigzag64()
  tss.push(firstTs)
  let prev = firstTs
  for (let i = 1; i < numEvents; i++) {
    const delta = r.readLEB128Big()
    if (delta < 0n) throw new Error("non_monotone_timestamp: ts delta must be ≥0")
    prev = prev + BigInt(delta)
    tss.push(prev)
  }
  return tss
}

// ── Public: snapshot frame (event batch) ──────────────────────────────────
//
// Returns { schemaHash: Uint8Array, seqs: BigInt[], tss: BigInt[], columns: [...] }.

export function decodeBatch(bytes) {
  const r = new ByteReader(bytes)
  const flag = r.readByte()
  if (flag !== FRAME_SNAPSHOT) {
    if (flag === FRAME_DELTA) throw new Error("expected event batch (0x00), got delta chain (0x01)")
    if (flag === FRAME_STREAM_HEADER) throw new Error("expected event batch (0x00), got stream header (0x02)")
    throw new Error(`unknown frame flag 0x${flag.toString(16)}`)
  }

  const schemaHash = r.readBytes(SCHEMA_HASH_BYTES)
  const numEvents  = Number(r.readLEB128Big())

  const seqs = readSeqBlock(r, numEvents)
  const tss  = readTsBlock(r, numEvents)

  const numUserCols = r.readLEB128()
  const columns = []
  for (let c = 0; c < numUserCols; c++) {
    columns.push(readColumnBlock(r, numEvents))
  }

  return { schemaHash, seqs, tss, columns }
}

// ── Public: stream header ─────────────────────────────────────────────────
//
// Returns { streamId: Uint8Array, source: string, schemaHash: Uint8Array, seqStart: BigInt }.

export function decodeStreamHeader(bytes) {
  const r = new ByteReader(bytes)
  const flag = r.readByte()
  if (flag !== FRAME_STREAM_HEADER)
    throw new Error(`expected stream header (0x02), got 0x${flag.toString(16)}`)

  const streamId    = r.readBytes(STREAM_ID_BYTES)
  const sourceLen   = r.readLEB128()
  const sourceBytes = r.readBytes(sourceLen)
  let source
  try { source = _dec.decode(sourceBytes) }
  catch { throw new Error("invalid_utf8: source contains invalid UTF-8") }

  const schemaHash = r.readBytes(SCHEMA_HASH_BYTES)
  const seqStart   = r.readLEB128Big()

  return { streamId, source, schemaHash, seqStart }
}

// ── Op decoding ────────────────────────────────────────────────────────────

function readEventAppend(r) {
  const numEvents = Number(r.readLEB128Big())
  const seqs = readSeqBlock(r, numEvents)
  const tss  = readTsBlock(r, numEvents)
  const numCols = r.readLEB128()
  const columns = []
  for (let c = 0; c < numCols; c++) columns.push(readColumnBlock(r, numEvents))
  return { op: OP.EVENT_APPEND, seqs, tss, columns }
}

function readFieldUpdate(r) {
  const seq = r.readLEB128Big()
  const numCols = r.readLEB128()
  const columns = []
  for (let c = 0; c < numCols; c++) {
    const colId    = r.readLEB128()
    const typeByte = r.readByte()
    const ctype    = typeByte & 0x1F
    const hasValue = ((typeByte >> 5) & 1) === 1
    if (ctype > 16) throw new Error(`unknown_ctype: ctype ${ctype} ≥ 17 is reserved`)
    const value = hasValue ? readValue(r, ctype) : null
    columns.push({ colId, ctype, hasValue, value })
  }
  return { op: OP.FIELD_UPDATE, seq, columns }
}

function readEventExpire(r) {
  const seqLo = r.readLEB128Big()
  const seqHi = r.readLEB128Big()
  if (seqLo > seqHi) throw new Error(`invalid_seq_range: seq_lo (${seqLo}) > seq_hi (${seqHi})`)
  return { op: OP.EVENT_EXPIRE, seqLo, seqHi }
}

function readSchemaEvolve(r) {
  const subOp = r.readByte()
  if (subOp > 2) throw new Error(`unknown_schema_sub_op: sub_op ${subOp} is reserved`)
  switch (subOp) {
    case SCHEMA_SUB_OP.COLUMN_ADD: {
      const colId    = r.readLEB128()
      const typeByte = r.readByte()
      const ctype    = typeByte & 0x1F
      const nullable = ((typeByte >> 5) & 1) === 1
      if (ctype > 16) throw new Error(`unknown_ctype: ctype ${ctype} ≥ 17 is reserved`)
      const nameLen = r.readLEB128()
      if (nameLen === 0) throw new Error("invalid_col_name: empty name")
      const nameBytes = r.readBytes(nameLen)
      const name = _dec.decode(nameBytes)
      return { op: OP.SCHEMA_EVOLVE, subOp, colId, ctype, nullable, name }
    }
    case SCHEMA_SUB_OP.COLUMN_DROP: {
      const colId = r.readLEB128()
      return { op: OP.SCHEMA_EVOLVE, subOp, colId }
    }
    case SCHEMA_SUB_OP.COLUMN_RENAME: {
      const colId = r.readLEB128()
      const nameLen = r.readLEB128()
      if (nameLen === 0) throw new Error("invalid_col_name: empty name")
      const nameBytes = r.readBytes(nameLen)
      const name = _dec.decode(nameBytes)
      return { op: OP.SCHEMA_EVOLVE, subOp, colId, name }
    }
  }
}

function readCursorCheckpoint(r) {
  const seq = r.readLEB128Big()
  const nameLen = r.readLEB128()
  if (nameLen === 0) throw new Error("invalid_cursor_name: empty cursor name")
  const nameBytes = r.readBytes(nameLen)
  const name = _dec.decode(nameBytes)
  return { op: OP.CURSOR_CHECKPOINT, seq, name }
}

function readOp(r) {
  const opCode = r.readByte()
  switch (opCode) {
    case OP.EVENT_APPEND:      return readEventAppend(r)
    case OP.FIELD_UPDATE:      return readFieldUpdate(r)
    case OP.EVENT_EXPIRE:      return readEventExpire(r)
    case OP.SCHEMA_EVOLVE:     return readSchemaEvolve(r)
    case OP.CURSOR_CHECKPOINT: return readCursorCheckpoint(r)
    default:
      throw new Error(`unknown_delta_op: op code ${opCode} is reserved`)
  }
}

// ── Public: delta chain ────────────────────────────────────────────────────
//
// Returns { schemaHash: Uint8Array, ops: [...] }.

export function decodeChain(bytes) {
  const r = new ByteReader(bytes)
  const flag = r.readByte()
  if (flag !== FRAME_DELTA)
    throw new Error(`expected delta chain (0x01), got 0x${flag.toString(16)}`)

  const schemaHash = r.readBytes(SCHEMA_HASH_BYTES)
  const numOps = r.readLEB128()
  const ops = []
  for (let i = 0; i < numOps; i++) ops.push(readOp(r))
  return { schemaHash, ops }
}
